modules/logger/parser/failed_test_marker.go

```diff
@@ -0,0 +1,64 @@
+// Package logger/parser contains methods to parse and restructure log output from go testing and terratest
+package parser
+
+// TestResultMarker tracks the indentation level of a test result line in go test output.
+// Example:
+// --- FAIL: TestSnafu
+//     --- PASS: TestSnafu/Situation
+//     --- FAIL: TestSnafu/Normal
+// The three markers for the above in order are:
+// TestResultMarker{TestName: "TestSnafu", IndentLevel: 0}
+// TestResultMarker{TestName: "TestSnafu/Situation", IndentLevel: 4}
+// TestResultMarker{TestName: "TestSnafu/Normal", IndentLevel: 4}
+type TestResultMarker struct {
```

Comment:
I thought about this. Piecing out the terratest logs in the core lib are easy enough (since WE control when those are written). However, extracting the go testing logs for detecting pass/fail is hard. Specifically, I have yet to find a good way to get hooks into the test itself so that you could intercept the test entries. Without that, you would only get the terratest logs but not the result entries (still useful). The lack of hooks argument is (sort of) supported by the community implementations of junit reporting. I found two implementations, [one for gradle](https://github.com/echocat/gradle-golang-plugin/commit/8464789852c305fb36ad5d4a3cba03bc13ddb046#diff-94b9a204193aa8994563f6f4e6f93474R1), and the other is [a command, which I am using here](https://github.com/jstemmer/go-junit-report), both of which rely on after the fact parsing to get the results out rather than baking it in to `go test` as a plugin.

The best alternative way I found so far for capturing test results is redefining `T`, but that would require changing all tests to use the new `T` type instead of go's `T` type. This also requires providing a custom `TestRunner` so that it will build the new `T` type. At this point, we are basically reimplementing the testing infrastructure of go, and not sure if that is good.

Given that assumption (that we can't grab `go test` logs on the fly), here are some additional bullet list pros and cons for the current approach that I came up with to further inform the decision:

Pros:
- Robust to unexpected failures in testing code, like `ctrl+c`, panics, OOM kills and the like since the parser is not tied to the testing process. This approach is less likely to miss these entries, and can be surfaced to the summary view for easy viewing in CI engines (no need to scroll), like the panic example.
- Can combine `go test` output (e.g `--- PASS` entries) with the log entries for the test in a single log file.
- Can extract out the summary view (those are all `go test` log entries).
- Can build `junit.xml` report that CI engines can use for test insights.

Cons:
- Complicated implementation that is potentially brittle. E.g if someone decides to change the logging format then this will all break. If we hook during the test, then the implementation is easier because those logs are all emitted at certain points in code, the information of which is lost in the final log and have to parse out.
- Have to store all the logs twice (the full interleaved version, and the broken out version) because the parsing depends on logs being available. (NOTE: this is avoidable with a pipe, but some CI engines are not good with pipes. For example, I couldn't get the piped version to work properly in circleCI).

Reply:
That's a good analysis. You should probably add it to the docs! There doesn't seem to be a perfect solution here, so let's go with what you have, as it's 100x better than what we had before. Thanks!

Reply:
That's a good analysis. You should probably add it to the docs! There doesn't seem to be a perfect solution here, so let's go with what you have, as it's 100x better than what we had before. Thanks!

Reply:
I would like to +1 the sentiment that it can be improved at a later point